src/userbase-server/purge.js

```diff
@@ -0,0 +1,461 @@
+
+  await Promise.all([
+    purgeAccessTokens(admin),
+    purgeApps(admin)
+  ])
+
+  // await connection.ddbClient().delete({
+  //   TableName: setup.deletedAdminsTableName,
+  //   Key: {
+  //     'admin-id': admin['admin-id']
+  //   }
+  // }).promise()
+  logger.child(logChildObject).info('Log placeholder -- Deleting Admin')
+
+  logger.child({ timeToPurge: Date.now() - start, ...logChildObject }).info('Finished purging admin')
+}
+
+const purgeDeletedUsers = async (purgeId) => {
+  const start = Date.now()
+  const logChildObject = { purgeId }
+  logger.child(logChildObject).info('Purging deleted users')
+
+  const params = {
+    TableName: setup.deletedUsersTableName
+  }
+
+  const ddbClient = connection.ddbClient()
+
+  const ddbQuery = (params) => ddbClient.scan(params).promise()
+  const action = (users) => Promise.all(users.map(user => purgeUser(user)))
+  await ddbWhileLoop(params, ddbQuery, action)
+
+  logger.child({ timeToPurge: Date.now() - start, ...logChildObject }).info('Finished purging deleted users')
+}
+
+const purgeDeletedApps = async (purgeId) => {
+  const start = Date.now()
+  const logChildObject = { purgeId }
+  logger.child(logChildObject).info('Purging deleted apps')
+
+  const params = {
+    TableName: setup.deletedAppsTableName
+  }
+
+  const ddbClient = connection.ddbClient()
+
+  const ddbQuery = (params) => ddbClient.scan(params).promise()
+  const action = (apps) => Promise.all(apps.map(app => purgeApp(app)))
+  await ddbWhileLoop(params, ddbQuery, action)
+
+  logger.child({ timeToPurge: Date.now() - start, ...logChildObject }).info('Finished purging deleted apps')
+}
+
+const purgeDeletedAdmins = async (purgeId) => {
+  const start = Date.now()
+  const logChildObject = { purgeId }
+  logger.child(logChildObject).info('Purging deleted admins')
+
+  const params = {
+    TableName: setup.deletedAdminsTableName
+  }
+
+  const ddbQuery = (params) => connection.ddbClient().scan(params).promise()
+  const action = (admins) => Promise.all(admins.map(admin => purgeAdmin(admin)))
+  await ddbWhileLoop(params, ddbQuery, action)
+
+  logger.child({ timeToPurge: Date.now() - start, ...logChildObject }).info('Finished purging deleted admins')
+}
+
+const commencePurge = async (purgeId) => {
+  const start = Date.now()
+  const logChildObject = { purgeId }
+
+  try {
+    logger.child(logChildObject).info('Commencing purge')
+
+    // place deleted items in permanent deleted tables
+    await Promise.all([
+      scanForDeletedAdmins(purgeId),
+      scanForDeletedApps(purgeId),
+      scanForDeletedUsers(purgeId),
+    ])
+
+    // purge items from permanent deleted tables. Do each synchronously because top level may delete level below it;
+    // for example, purging admins will purge apps and users, reducing the number of deleted apps and deleted users
+    await purgeDeletedAdmins(purgeId)
+    await purgeDeletedApps(purgeId)
+    await purgeDeletedUsers(purgeId)
```

Comment:
Purging an admin or app is the full purge process that goes from the bottom up such that there won't be orphaned items.

When it purges an app for example, it will first iterate over all the app's users, then delete each user individually, then delete the app.

When it purges an admin, it will first iterate over all the admin's apps, then all that app's users, then delete each user individually, then delete the apps, then delete the admin.

I was thinking going in this order would improve efficiency, but after thinking more on it, that won't really be the case until we get to database sharing. I'm inclined to keep it for that reason. Getting into the details a little, purging a single user will have slightly more complex logic as it relates to how databases are purged when multiple users share the database (it will need to do an additional query to see if other users share the database). But deleting an app or admin can simply get rid of the need for doing that extra logic.